test/liquidityConversionRate.js

```diff
@@ -869,6 +879,211 @@ contract('kyberReserve for Liquidity', function(accounts) {
+
+                    assert(expectedDest * 1.0 <= totalAmount * 1.0, "expected dest amount should be lower than previous traded amount, loop: " + i);
+                }
+            }
+        }
+    });
+
+    // init new conversion rate, reserve, make a buy then check sell rate, make a sell and check buy rate
+    it("Should allow buy/sell all inventory, check rate never fall above max rate or below min rate", async function() {
+        // if i < newRs.length, we test buy then check rate
+        // otherwise test sell then check rate
+        for (let i = 0; i < newRs.length * 2; i++) {
+            let id = i < newRs.length ? i : i - newRs.length;
+            let currentBlock = await Helper.getCurrentBlock();
+
+            let token = await TestToken.new("test", "tst", 18);
+
+            let newLiqConvRatesInst = await LiquidityConversionRates.new(admin, token.address);
+
+            let feeInBps = 0;
+            let pMin = newP0s[id] * newpMinRatios[id];
+            let pMax = newP0s[id] * newpMaxRatios[id];
+
+            let rInFp = BigNumber(newRs[id]).mul(formulaPrecision);
+            let pMinInFp = BigNumber(pMin).mul(formulaPrecision);
+            let maxCapBuyInWei = BigNumber(newMaxCapBuyInEth[id]).mul(precision);
+            let maxCapSellInWei = BigNumber(newMaxCapSellInEth[id]).mul(precision);
+            let maxSellRateInPrecision = BigNumber(pMax).mul(precision);
+            let minSellRateInPrecision = BigNumber(pMin).mul(precision);
+
+            await newLiqConvRatesInst.setLiquidityParams(
+                rInFp,
+                pMinInFp,
+                formulaPrecisionBits,
+                maxCapBuyInWei,
+                maxCapSellInWei,
+                feeInBps,
+                maxSellRateInPrecision,
+                minSellRateInPrecision
+            );
+
+            let reserveInst = await Reserve.new(network, newLiqConvRatesInst.address, admin);
+            await reserveInst.setContracts(network, newLiqConvRatesInst.address, 0);
+
+            await newLiqConvRatesInst.setReserveAddress(reserveInst.address);
+
+            //set reserve balance.
+            let reserveEtherInit = (BigNumber(10).pow(18)).mul(newE0s[id]);
+            await Helper.sendEtherWithPromise(accounts[9], reserveInst.address, reserveEtherInit);
+
+            await reserveInst.approveWithdrawAddress(token.address, accounts[0], true);
+
+            //transfer tokens to reserve.
+            let amount = (BigNumber(10).pow(18)).mul(newT0s[id]);
+            await token.transfer(reserveInst.address, amount.valueOf());
+
+            if (i < newRs.length) {
+                // buy then check sell rate
+                let lastBuyRate = BigNumber(precision).div(newP0s[id]).floor();
+                let srcAmount = maxCapBuyInWei;
+                while (true) {
+                    let bal = await token.balanceOf(reserveInst.address);
+                    if (bal * 1.0 <= precision * 1.0) { break; }
```

Comment:
I don't understand how balance is related to precision?
is it because token has 18  decimals?